src/link/object.cpp

```diff
@@ -599,33 +604,33 @@ void obj_ReadFile(char const *fileName, unsigned int fileID) {
 	// Give symbols' section pointers to their sections
 	for (uint32_t i = 0; i < nbSymbols; i++) {
 		if (Label *label = std::get_if<Label>(&fileSymbols[i].data); label) {
-			Section *section = fileSections[label->sectionID];
+			Section *section = fileSections[label->sectionID].get();
 
+			label->section = section;
 			// Give the section a pointer to the symbol as well
 			linkSymToSect(fileSymbols[i], *section);
+		}
+	}
 
-			if (section->modifier != SECTION_NORMAL) {
+	// Calling `sect_AddSection` invalidates the contents of `fileSections`!
+	for (uint32_t i = 0; i < nbSections; i++)
+		sect_AddSection(std::move(fileSections[i]));
+
+	// Fix symbols' section pointers to component sections
+	// This has to run **after** all the `sect_AddSection()` calls,
+	// so that `sect_GetSection()` will work
+	for (uint32_t i = 0; i < nbSymbols; i++) {
+		if (Label *label = std::get_if<Label>(&fileSymbols[i].data); label) {
+			if (Section *section = label->section; section->modifier != SECTION_NORMAL) {
 				if (section->modifier == SECTION_FRAGMENT)
 					// Add the fragment's offset to the symbol's
 					label->offset += section->offset;
-				section = getMainSection(*section);
+				// Associate the symbol with the main section, not the "component" one
+				label->section = sect_GetSection(section->name);
```

Comment:
Not sure it's great to always do a lookup when `section` is correct most of the time, but eh.

Reply:
We already were. The `getMainSection` helper I deleted was doing a `section.modifier != SECTION_NORMAL` check, but it was redundant because this whole block only runs if that's the case.